File: apps/ios/Pebbles/Features/Lab/Services/LogsService.swift
import Foundation
import Supabase
import os

/// Read/write access to `public.logs` and `public.log_reactions` for the Lab tab.
/// Reads go through `v_logs_with_counts` so every row carries its reaction count.
/// Reactions are single-table writes on `log_reactions` — no RPC needed.
@MainActor
struct LogsService {
    let supabase: SupabaseService

    private static let logger = Logger(subsystem: "app.pbbls.ios", category: "logs-service")

    // MARK: - Feeds

    /// Published announcements, most recent first.
    func announcements(limit: Int? = nil) async throws -> [Log] {
        let base = supabase.client
            .from("v_logs_with_counts")
            .select()
            .eq("species", value: LogSpecies.announcement.rawValue)
            .eq("published", value: true)
            .order("published_at", ascending: false)
        if let limit {
            let wrapper: LossyLogArray = try await base.limit(limit).execute().value
            return wrapper.logs
        }
        let wrapper: LossyLogArray = try await base.execute().value
        return wrapper.logs
    }

    /// Shipped features, most recent first.
    func changelog(limit: Int? = nil) async throws -> [Log] {
        let base = supabase.client
            .from("v_logs_with_counts")
            .select()
            .eq("species", value: LogSpecies.feature.rawValue)
            .eq("status", value: LogStatus.shipped.rawValue)
            .eq("published", value: true)
            .order("published_at", ascending: false)
        if let limit {
            let wrapper: LossyLogArray = try await base.limit(limit).execute().value
            return wrapper.logs
        }
        let wrapper: LossyLogArray = try await base.execute().value
        return wrapper.logs
    }

    /// Features currently in progress.
    func initiatives() async throws -> [Log] {
        let wrapper: LossyLogArray = try await supabase.client
            .from("v_logs_with_counts")
            .select()
            .eq("species", value: LogSpecies.feature.rawValue)
            .eq("status", value: LogStatus.inProgress.rawValue)
            .eq("published", value: true)
            .order("published_at", ascending: false)
            .execute()
            .value
        return wrapper.logs
    }

    /// Backlog features, most upvoted first. Ties broken by recency.
    func backlog(limit: Int? = nil) async throws -> [Log] {
        let base = supabase.client
            .from("v_logs_with_counts")
            .select()
            .eq("species", value: LogSpecies.feature.rawValue)
            .eq("status", value: LogStatus.backlog.rawValue)
            .eq("published", value: true)
            .order("reaction_count", ascending: false)
            .order("created_at", ascending: false)
        if let limit {
            let wrapper: LossyLogArray = try await base.limit(limit).execute().value
            return wrapper.logs
        }
        let wrapper: LossyLogArray = try await base.execute().value
        return wrapper.logs
    }

    // MARK: - Reactions

    /// The set of log ids the current user has upvoted. Used by views to
    /// render the reaction button in its "reacted" state and to pick the
    /// right toggle action. Returns an empty set if no session.
    func myReactions() async throws -> Set<UUID> {
        guard let userId = supabase.session?.user.id else { return [] }
        let rows: [ReactionRow] = try await supabase.client
            .from("log_reactions")
            .select("log_id")
            .eq("user_id", value: userId)
            .execute()
            .value
        return Set(rows.map(\.logId))
    }

    /// Upvote a log on behalf of the current user. No-op on unique-key conflict —
    /// the PK `(log_id, user_id)` ensures each user can react at most once.
    func react(logId: UUID) async throws {
        guard let userId = supabase.session?.user.id else {
            throw LogsServiceError.missingSession
        }
        do {
            try await supabase.client
                .from("log_reactions")
                .insert(ReactionInsert(logId: logId, userId: userId))
                .execute()
        } catch {
            Self.logger.error("react failed: \(error.localizedDescription, privacy: .private)")
            throw error
        }
    }

    /// Remove the current user's upvote on a log.
    func unreact(logId: UUID) async throws {
        guard let userId = supabase.session?.user.id else {
            throw LogsServiceError.missingSession
        }
        do {
            try await supabase.client
                .from("log_reactions")
                .delete()
                .eq("log_id", value: logId)
                .eq("user_id", value: userId)
                .execute()
        } catch {
            Self.logger.error("unreact failed: \(error.localizedDescription, privacy: .private)")
            throw error
        }
    }
}

// MARK: - Wire types

private struct ReactionRow: Decodable {
    let logId: UUID
    enum CodingKeys: String, CodingKey { case logId = "log_id" }
}

private struct ReactionInsert: Encodable {
    let logId: UUID
    let userId: UUID
    enum CodingKeys: String, CodingKey {
        case logId = "log_id"
        case userId = "user_id"
    }
}

enum LogsServiceError: Error, LocalizedError {
    case missingSession

    var errorDescription: String? {
        switch self {
        case .missingSession: return "Please sign in again."
        }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/core/data/LogsService.kt
package app.pbbls.android.core.data

import app.pbbls.android.AppEnvironment
import app.pbbls.android.core.model.LabConfig
import app.pbbls.android.core.model.Log
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.postgrest.query.PostgrestRequestBuilder
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.OffsetDateTime
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "logs-service"

/**
 * The Lab seam (#849) — what [app.pbbls.android.features.lab.LabViewModel] and
 * [app.pbbls.android.features.lab.LogListViewModel] are tested against.
 */
interface LogsServicing {
    suspend fun announcements(limit: Int? = null): List<Log>

    /**
     * One row by [id], or null if it does not exist. A single-table,
     * single-statement read, so no RPC (root `AGENTS.md`) — the seam
     * [AnnouncementDetailViewModel] loads by id, mirroring how
     * [app.pbbls.android.features.profile.SoulDetailViewModel] fetches its
     * soul rather than receiving the row from whichever screen had it.
     */
    suspend fun log(id: String): Log?

    suspend fun changelog(limit: Int? = null): List<Log>

    suspend fun initiatives(): List<Log>

    suspend fun backlog(limit: Int? = null): List<Log>

    suspend fun myReactions(): Set<String>

    suspend fun react(logId: String)

    suspend fun unreact(logId: String)

    /**
     * Not a call — a pure projection of [Log.coverImagePath] against the project
     * URL. On the interface so the ViewModel can fold it into state and the
     * screen never reads a service; a fake supplies its own base rather than
     * touching `AppEnvironment`.
     */
    fun coverImageUrl(log: Log): String?
}

/**
 * The Lab data layer — ports iOS `LogsService` (M44 design D3/D4). Four feed
 * reads over `v_logs_with_counts` (decoded lossily via [LossyLogList]) and the
 * reaction writes, which go straight to `log_reactions` — a single-table,
 * single-statement insert/delete on PK `(log_id, user_id)`, the sanctioned
 * direct-write pattern (deliberately no RPC, per root `AGENTS.md`; iOS
 * comments the same). Methods throw; screens own view state and the
 * optimistic revert.
 */

@Singleton
class LogsService
    @Inject
    constructor(
        private val supabase: SupabaseService,
    ) : LogsServicing {
        /** Published announcements, newest published first. */
        override suspend fun announcements(limit: Int?): List<Log> =
            feed {
                filter {
                    eq("species", "announcement")
                    eq("published", true)
                }
                order("published_at", Order.DESCENDING)
                limit?.let { limit(it.toLong()) }
            }

        /** One row by [id], undecorated by any species/status filter — see the interface doc. */
        override suspend fun log(id: String): Log? =
            supabase.client
                .from("v_logs_with_counts")
                .select {
                    filter { eq("id", id) }
                }.decodeSingleOrNull<Log>()

        /**
         * Shipped features — `released_at` DESC with nulls LAST (the easy-to-miss
         * D3 detail), then `published_at` DESC. The client-side re-sort below
         * guarantees the rendered order matches iOS even if the query builder
         * collapses the two `order` calls; only a `limit` truncation under tied
         * timestamps could theoretically differ (design risk 3).
         */
        override suspend fun changelog(limit: Int?): List<Log> =
            feed {
                filter {
                    eq("species", "feature")
                    eq("status", "shipped")
                    eq("published", true)
                }
                order("released_at", Order.DESCENDING, nullsFirst = false)
                order("published_at", Order.DESCENDING)
                limit?.let { limit(it.toLong()) }
            }.sortedWith(changelogOrder)

        /** Features in progress, newest published first — always unlimited (iOS). */
        override suspend fun initiatives(): List<Log> =
            feed {
                filter {
                    eq("species", "feature")
                    eq("status", "in_progress")
                    eq("published", true)
                }
                order("published_at", Order.DESCENDING)
            }

        /** Backlog features — most upvoted first, then newest created. */
        override suspend fun backlog(limit: Int?): List<Log> =
            feed {
                filter {
                    eq("species", "feature")
                    eq("status", "backlog")
                    eq("published", true)
                }
                order("reaction_count", Order.DESCENDING)
                order("created_at", Order.DESCENDING)
                limit?.let { limit(it.toLong()) }
            }.sortedWith(backlogOrder)

        /**
         * The caller's reacted log ids. No session → empty set WITHOUT throwing
         * (iOS: an anonymous Lab renders with zero reactions); the writes below
         * DO throw without a session.
         */
        override suspend fun myReactions(): Set<String> {
            val me = supabase.session?.user?.id ?: return emptySet()
            return supabase.client
                .from("log_reactions")
                .select(Columns.list("log_id")) {
                    filter { eq("user_id", me) }
                }.decodeList<ReactionRow>()
                .map { it.logId }
                .toSet()
        }

        /**
         * Upvote — inserts the `(log_id, user_id)` row. A duplicate insert
         * conflicts on the PK and throws like any failure, triggering the
         * caller's optimistic revert.
         */
        override suspend fun react(logId: String) {
            val me = requireUserId()
            try {
                supabase.client
                    .from("log_reactions")
                    .insert(ReactionWrite(logId = logId, userId = me))
            } catch (e: Exception) {
                android.util.Log.e(TAG, "react failed", e)
                throw e
            }
        }

        /** Remove an upvote — deletes by both PK columns. */
        override suspend fun unreact(logId: String) {
            val me = requireUserId()
            try {
                supabase.client
                    .from("log_reactions")
                    .delete {
                        filter {
                            eq("log_id", logId)
                            eq("user_id", me)
                        }
                    }
            } catch (e: Exception) {
                android.util.Log.e(TAG, "unreact failed", e)
                throw e
            }
        }

        /** Public cover-image URL for [log], or null (design D7). */
        override fun coverImageUrl(log: Log): String? = LabConfig.coverImageUrl(AppEnvironment.supabaseUrl, log.coverImagePath)

        // android.util.Log stays fully qualified in this file — the imported Log
        // is the Lab model.
        private suspend fun feed(builder: PostgrestRequestBuilder.() -> Unit): List<Log> {
            val raw =
                supabase.client
                    .from("v_logs_with_counts")
                    .select(request = builder)
                    .data
            return LossyLogList.decode(raw) { index, cause ->
                android.util.Log.e(TAG, "skipped log row $index", cause)
            }
        }

        private fun requireUserId(): String =
            supabase.session?.user?.id
                ?: throw IllegalStateException("lab reaction without session")

        @Serializable
        private data class ReactionRow(
            @SerialName("log_id")
            val logId: String,
        )

        @Serializable
        private data class ReactionWrite(
            @SerialName("log_id")
            val logId: String,
            @SerialName("user_id")
            val userId: String,
        )

        companion object {
            /** iOS changelog order: released_at DESC nulls last, then published_at DESC. */
            internal val changelogOrder: Comparator<Log> =
                compareByDescending<Log> { it.releasedAt != null }
                    .thenByDescending { it.releasedAt ?: OffsetDateTime.MIN }
                    .thenByDescending { it.publishedAt ?: OffsetDateTime.MIN }

            /** iOS backlog order: reaction_count DESC, then created_at DESC. */
            internal val backlogOrder: Comparator<Log> =
                compareByDescending<Log> { it.reactionCount }
                    .thenByDescending { it.createdAt }
        }
    }
